backends/gstreamer/src/player.rs

```diff
@@ -375,23 +375,89 @@ impl GStreamerPlayer {
+            appsink.set_caps(&caps);
+
+            glsink
+                .set_property("sink", &appsink)
+                .map_err(|e| PlayerError::Backend(e.to_string()))?;
+
+            pipeline
+                .set_property("video-sink", &glsink.to_value())
+                .map_err(|e| PlayerError::Backend(e.to_string()))?;
+        } else {
+            let caps = gst::Caps::builder("video/x-raw")
+                .field("format", &gst_video::VideoFormat::Bgra.to_string())
+                .field("pixel-aspect-ratio", &gst::Fraction::from((1, 1)))
+                .build();
+            appsink.set_caps(&caps);
+
+            pipeline
+                .set_property("video-sink", &appsink.to_value())
+                .map_err(|e| PlayerError::Backend(e.to_string()))?;
+        };
 
-        pipeline
-            .set_property("video-sink", &video_sink.to_value())
-            .map_err(|e| PlayerError::Backend(e.to_string()))?;
+        if self.gl_context.borrow().is_some() {
+            use gst::ElementExt;
+
+            let bus = pipeline
+                .get_bus()
+                .ok_or(PlayerError::Backend("pipeline without bus".to_owned()))?;
+
+            let gl_context = self.gl_context.replace(None).unwrap();
+            let gl_display = self.gl_display.replace(None).unwrap();
+            bus.set_sync_handler(move |_, msg| {
+                use gst::MessageView;
+
+                match msg.view() {
+                    MessageView::NeedContext(ctxt) => {
+                        let context_type = ctxt.get_context_type();
+                        if context_type == *gst_gl::GL_DISPLAY_CONTEXT_TYPE {
+                            if let Some(el) =
+                                msg.get_src().map(|s| s.downcast::<gst::Element>().unwrap())
+                            {
+                                use gst_gl::ContextGLExt;
+
+                                let context = gst::Context::new(context_type, true);
+                                context.set_gl_display(&gl_display);
+                                el.set_context(&context);
+                            }
+                        }
+                        if context_type == "gst.gl.app_context" {
+                            if let Some(el) =
+                                msg.get_src().map(|s| s.downcast::<gst::Element>().unwrap())
+                            {
+                                let mut context = gst::Context::new(context_type, true);
+                                {
+                                    let context = context.get_mut().unwrap();
+                                    let mut s = context.get_mut_structure();
+                                    s.set_value("context", gl_context.to_send_value());
+                                }
+                                el.set_context(&context);
+                            }
+                        }
+                    }
+                    _ => (),
+                }
```

Comment:
A slightly simpler way of doing the same thing:

```rust
match msg.view() {
    MessageView::NeedContext(ctxt) => {
        if let Some(el) =
            msg.get_src().map(|s| s.downcast::<gst::Element>().unwrap())
        {
            use gst_gl::ContextGLExt;
            
            let context_type = ctxt.get_context_type();
            let mut context = gst::Context::new(context_type, true);
            if context_type == *gst_gl::GL_DISPLAY_CONTEXT_TYPE {
                context.set_gl_display(&gl_display);
            } else if context_type == "gst.gl.app_context" {
                let mut s = context.get_mut().unwrap().get_mut_structure();
                s.set_value("context", gl_context.to_send_value());
            };
            el.set_context(&context);
        }
    }
    _ => (),
}
```

Reply:
The problem there is we shouldn't set a void context (not display nor app)